### DISCUSSION TITLE
Suggestion for dealing with keybind conflicts in vscode terminal?

### DISCUSSION THREAD

**Discussion by digitalsignalperson** [akinomyoga/ble.sh#408] [Q&A] [answered]
For example C-j in multiline is the default keybind in vscode to open/close the terminal panel.

Do people here change all the keybinds in vscode, or in ble.sh, or is there some other trick maybe? Aside from not using vscode.

**Comment by akinomyoga** [comment]
> Do people here

The community hasn't grown up, so there are no people here.

When there are conflicts in the keybindings with the terminal, usually, the keybindings of either need to be changed.

I'm not familiar with VS Code Terminal, but after a quick search, the config `terminal.integrated.commandsToSkipShell` might be related. You can press <kbd>C-,</kbd> to open Settings and type `terminal.integrated.commandsToSkipShell` in the textbox to find the configuration. The description of the config says

> A set of command IDs whose keybindings will not be sent to the shell but instead always be handled by VS Code. This allows keybindings that would normally be consumed by the shell to act instead the same as when the terminal is not focused, for example Ctrl+P to launch Quick Open.
>
> Many commands are skipped by default. To override a default and pass that command's keybinding to the shell instead, add the command prefixed with the - character. For example add -workbench.action.quickOpen to allow Ctrl+P to reach the shell.
>
> The following list of default skipped commands is truncated when viewed in Settings Editor. To see the full list, *open the default settings JSON* and search for the first command from the list below.
>
> Default Skipped Commands:
>
> [...very long list...]

You need to pick up the command name that is bound to <kbd>C-j</kbd>, and add an item "`-<command name>`" to revert the skipping.

**Comment by akinomyoga** [comment]
I found a simpler config: `terminal.integrated.sendKeybindingsToShell`.

You can find it by pressing <kbd>C-,</kbd> (Control + comma) to open Settings and by typing `terminal.integrated.sendKeybindingsToShell` and just turn on the check box. This made it possible to input <kbd>C-j</kbd> in VS Code Terminal in my environment.

**Reply by akinomyoga** [reply, marked as answer]
The description of `terminal.integrated.sendKeybindingsToShell` says

> Dispatches most keybindings to the terminal instead of the workbench, overriding `terminal.integrated.commandsToSkipShell`, which can be used alternatively for fine tuning.

so if you need fine-tuning on which should be sent to the terminal and which shouldn't, you'll need to configure `terminal.integrated.commandsToSkipShell` mentioned in another reply.

**Reply by digitalsignalperson** [reply]
Thanks for those tips, that seems like a pretty complete set of possibilities. I'm used to C-j for opening/closing the vscode integrated terminal panel, so maybe I'll pick one to re-bind. Otherwise I'd have to think if set/unset that sendKeybindingsToShell on demand can be of any use.  For keybindings, I like the multi-line editing experience in the python qtconsole, so maybe I'll try copying those keybindings (Ctrl+Enter = new line, Shift+Enter = execute, or Enter = execute if on the last line)

Btw I'm also happy to be checking this project out. I heard ble.sh is the largest shell program in the world!  I was getting frustrated with fish as my main shell when all the scripting I'm doing is with bash. This is really cool to have a bash environment with practically all the nice bits from fish.

**Reply by akinomyoga** [reply]
> I'll try copying those keybindings (Ctrl+Enter = new line, Shift+Enter = execute, or Enter = execute if on the last line)

I actually tried to enable those key combinations, <kbd>Ctrl+Enter</kbd> and <kbd>Shift+Enter</kbd>, for VS Code Terminal using `workbench.action.terminal.sendSequence` in `keybindings.json`, ~~but somehow the VS Code Terminal ignores the settings and always sends just <kbd>Enter</kbd>.~~ (**Edit**: It turned out that to use <kbd>ctrl+enter</kbd> and <kbd>shift+enter</kbd>, I need to turn off `terminal.integrated.sendKeybindingsToShell`.) Here's my settings in `keybindings.json`.

```js
    { "key": "shift+enter", "command": "workbench.action.terminal.sendSequence", "args": { "text": "\u001b[27;2;13~" }, "when": "terminalFocus" },
    { "key": "ctrl+enter", "command": "workbench.action.terminal.sendSequence", "args": { "text": "\u001b[27;5;13~" }, "when": "terminalFocus" },
    { "key": "ctrl+shift+enter", "command": "workbench.action.terminal.sendSequence", "args": { "text": "\u001b[27;6;13~" }, "when": "terminalFocus" },
```

I'm not sure if it is possible to bind those key combinations for VS Code Terminal.

> I heard ble.sh is the largest shell program in the world!

Yeah, that's the estimate by Andy Chu, the author of the Oil shell. There is still some auto-generated large scripts like `./configure` for GNU coreutils, but for the hand-written shell scripts available publicly, maybe this would be the largest one. At least, I'm not aware of other scripts that has the same scale. However, I feel there might be a larger script privately used in a company. Anyway, thank you for trying it!

**Reply by akinomyoga** [reply]
I sorted out how to bind <kbd>C-RET</kbd>, etc. in the VSCode Terminal. I summarized it in the wiki page: [Visual Studio Code Terminal - Manual A2](https://github.com/akinomyoga/ble.sh/wiki/Manual-A2-Terminal-Settings#visual-studio-code-terminal).

**Reply by digitalsignalperson** [reply]
Thanks for the thorough wiki guide.

A question on the multiline bindings themselves,

```
-- MULTILINE -- (RET or C-m: insert a newline, C-j: run)
```

Is it possible to change those RET/C-m/C-j bindings?

e.g. I tried changing any `  ble-bind -f 'C-j' ` instances I could find but there was no effect.

**Reply by akinomyoga** [reply]
> Is it possible to change those RET/C-m/C-j bindings?

Yes.

> e.g. I tried changing any ` ble-bind -f 'C-j'` instances I could find but there was no effect.

There are many possibilities. For example, Visual Studio Code may process <kbd>C-j</kbd> before <kbd>C-j</kbd> reaches ble.sh. Or, as another possibility, you bind the key in a different keymap. Or maybe the arguments to `ble-bind` were wrong.

- Currently, what is the resulting behavior of pressing <kbd>C-j</kbd>?
- Where did you put those `ble-bind` settings?
- Could you provide an actual `ble-bind ...` that you tried?

**Reply by digitalsignalperson** [reply]
Currently  I'm completely outside vscode with Konsole for my terminal. Say I'm looking for the "C-j: run" binding for multiline mode.  I don't think any of these are the right one:  `ble-bind -P | grep C-j`
```
ble-bind -m 'isearch' -f 'C-j' 'isearch/accept-line'
ble-bind -m 'nsearch' -f 'C-j' 'nsearch/accept-line'
ble-bind -m 'emacs' -f 'C-j' 'accept-line'
ble-bind -m 'auto_complete' -f 'C-j' 'auto_complete/accept-line'
```

What I tried is simply in `out/ble.sh`, find and replace all instances of `C-j` with `C-RET`, then go to a new terminal, `bash`, then `source out/ble.sh`, and enter multiline mode still shows  `-- MULTILINE -- (RET or C-m: insert a newline, C-j: run)` and C-j is still run.

**Reply by akinomyoga** [reply]
> What I tried is simply in `out/ble.sh`, find and replace all instances of `C-j` with `C-RET`,

Although that's not an expected way of configuring it, the keybinding in the Emacs editing mode (which is the default when you don't specify `set -o vi`) is defined in `out/lib/keymap.emacs.sh` but not in `out/ble.sh`. You could have replaced `C-j` in `out/lib/keymap.emacs.sh`.

However, the expected way of removing the keybinding to <kbd>C-j</kbd> is to put the following line in `~/.blerc` or `~/.config/blesh/init.sh`:

```bash
# blerc

ble-bind -m emacs -f 'C-j' ''
```

Also, <kbd>C-RET</kbd> is already bound by `accept-line`, so there is no need to configure it on the ble.sh side. If it doesn't work, you need to configure <kbd>C-RET</kbd> on the Konsole side. To use <kbd>C-RET</kbd>, you need to configure both sides of the terminal  (Konsole) and the terminal application (ble.sh). The configuration in Konsole is explained in https://github.com/akinomyoga/ble.sh/issues/187#issuecomment-1107551319. The code for <kbd>C-RET</kbd> (that should be specified in the *Output* column in the Konsole configuration) is `\E[13;5u`. You can check [§3.6.4](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A73-Key-Binding#user-content-sec-modifyOtherKeys-manual) for the codes of more key combinations.
